### DISCUSSION TITLE
C# Source Generator for DirectProperty registration...

### DISCUSSION THREAD

**Discussion by BenMakesGames** [AvaloniaUI/Avalonia#6315] [General]
existing Fody solutions for WPF dependent property boilerplate don't work, because AvaloniaUI rolled its property stuff.

I discovered C# Source Generators today, and hacked together... something? I've so far tested only on a single property in a single class; I have no doubt it has issues I haven't discovered yet, but here's what I've got so far...

example use:

```
public partial class MyUserControl : UserControl
{
    [RegisterDirectProperty] // dependent property boilerplate can gtfo!
    public int BorderThickness { get; set; }

    ...
}
```

attribute:

```
using System;

namespace Avalonia.CodeGenerators.Attributes
{
    [AttributeUsage(AttributeTargets.Property)]
    public class RegisterDirectPropertyAttribute : Attribute
    {
    }
}
```

code generator:

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Avalonia.CodeGenerators
{
    [Generator]
    public class DirectPropertyGenerator: ISourceGenerator
    {
        public const string REGISTER_DIRECT_PROPERTY_ATTRIBUTE_NAME = "Avalonia.CodeGenerators.Attributes.RegisterDirectPropertyAttribute";
        public const string AVALONIA_CONTROL_CLASS_NAME = "Avalonia.Controls.Control";

        public void Initialize(GeneratorInitializationContext context)
        {
/*
#if DEBUG
            if (!Debugger.IsAttached)
            {
                Debugger.Launch();
            }
#endif 
*/

            context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
        }

        public void Execute(GeneratorExecutionContext context)
        {
            if (!(context.SyntaxContextReceiver is SyntaxReceiver receiver))
                return;

            var attributeSymbol = context.Compilation.GetTypeByMetadataName(REGISTER_DIRECT_PROPERTY_ATTRIBUTE_NAME);
            var avaloniaControlSymbol = context.Compilation.GetTypeByMetadataName(AVALONIA_CONTROL_CLASS_NAME);

            foreach(var group in receiver.Properties.GroupBy(f => f.ContainingType))
            {
                var source = GenerateDirectPropertyRegistrationClassSource(group.Key, group.ToList(), attributeSymbol, avaloniaControlSymbol, context);
                context.AddSource($@"{group.Key.Name}.DirectProperties.cs", source);
            }
        }

        public string GenerateDirectPropertyRegistrationClassSource(INamedTypeSymbol classSymbol, List<IPropertySymbol> properties, ISymbol attributeSymbol, ISymbol avaloniaControlSymbol, GeneratorExecutionContext context)
        {
            var namespaceName = classSymbol.ContainingNamespace.ToDisplayString();
            var className = classSymbol.Name;

            var propertyRegistrations = properties.Select(f => GenerateDirectPropertyRegistration(className, f.Name, f.Type.ToString()));

            return $@"
using Avalonia;

namespace {namespaceName}
{{
    public partial class {className}
    {{
        {string.Join("\r\n", propertyRegistrations)}
    }}
}}
";
        }

        public string GenerateDirectPropertyRegistration(string className, string propertyName, string propertyType)
        {
            var directPropertyName = $"{propertyName}Property";

            return $@"
        public static readonly DirectProperty<{className}, {propertyType}> {directPropertyName} =
            AvaloniaProperty.RegisterDirect<{className}, {propertyType}>(
                nameof({propertyName}),
                o => o.{propertyName},
                (o, v) =>
                {{
                    var current = o.{propertyName};
                    o.SetAndRaise({directPropertyName}, ref current, v);
                    if (o.{propertyName} != current)
                        o.On{propertyName}Changed(o.{propertyName}, current);
                    o.{propertyName} = current;
                }}
            )
        ;
";
        }

        class SyntaxReceiver : ISyntaxContextReceiver
        {
            public List<IPropertySymbol> Properties { get; } = new List<IPropertySymbol>();

            /// <summary>
            /// Called for every syntax node in the compilation, we can inspect the nodes and save any information useful for generation
            /// </summary>
            public void OnVisitSyntaxNode(GeneratorSyntaxContext context)
            {
                // any field with at least one attribute is a candidate for property generation
                if (context.Node is PropertyDeclarationSyntax propertyDeclaration && propertyDeclaration.AttributeLists.Count > 0)
                {
                    // Get the symbol being declared by the field, and keep it if its annotated
                    var propertySymbol = context.SemanticModel.GetDeclaredSymbol(propertyDeclaration);

                    if (propertySymbol.GetAttributes().Any(ad => ad.AttributeClass.ToDisplayString() == REGISTER_DIRECT_PROPERTY_ATTRIBUTE_NAME))
                    {
                        Properties.Add(propertySymbol);
                    }
                }
            }
        }
    }
}
```

(if someone's already put together something way better than this, I'll feel _slightly_ bad about the time spent...)

I do feel like writing code generators is a band-aid. I think AvaloniaUI could do better: https://github.com/AvaloniaUI/Avalonia/discussions/6312 but until/unless that happens, code generators it is!

**Comment by jp2masa** [comment]
I created a source generator a few months ago ([Avalonia.PropertyGenerator](https://github.com/jp2masa/Avalonia.PropertyGenerator)) but instead of registering the CLR property, you have to register the Avalonia property (I decided it would be more flexible that way as the Avalonia properties registration has some additional metadata and it would be less lines of code on the original code).

Also, I think your implementation has a small mistake: the `SetAndRaise` has to be called in the CLR property setter, because if you try to update the CLR property from code it won't notify that the property changed.

**Reply by BenMakesGames** [reply]
I see what you're saying about my use of `SetAndRaise`. I tried to be _too clever_.

I'm very used to other front-end frameworks (Angular, Unity) where you just define - and maybe decorate - a single property, and it's good to go, so I'm trying to reproduce that ease-of-use.

I'll check out your code, for sure! thanks!

**Comment by maxkatz6** [comment]
Unfortunately, C# does not support partial properties yet. Not sure if it will in C#10. 
But without it we can't generate anything for your code example.

Potentially with partial keyword it would be possible to write C# code:
```
[RegisterDirectProperty]
public partial int BorderThickness { get; set; }
```
And auto generate:
```
public static DirectProperty BorderThicknessProperty = ...;

private int _borderThickness;
public partial int BorderThickness
{
   get => _borderThickness;
   set => this.SetAndRaise(BorderThicknessProperty, ref _borderThickness, value)
}
```

**Reply by maxkatz6** [reply]
Related discussion https://github.com/dotnet/csharplang/discussions/3412

**Reply by maxkatz6** [reply]
Although with new "field" modifier (which was confirmed to be in C#10) I can imagine it be a compromise solution:
```
[RegisterDirectProperty]
public partial int BorderThickness
{
   get => field;
   set => this.SetAndRaise(BorderThicknessProperty, ref field, value)
}
```
would have the same code generated. Of course, in this approach source will rely on static field (BorderThicknessProperty) that might be or might not be already generated. With Visual Studio it should be acceptable, it's quite stress-less with source generators now, but with Rider it will be very annoying, it's still bad with generating code.

As for me, it's still better than having attribute on field and generating property. Because in code we usually use them as properties and not fields.

**Reply by maxkatz6** [reply]
Anyway, I can't imagine something like this being include in-box or used inside of Avalonia repo in near future.
But I see potential of it being an optional feature.

**Comment by maxkatz6** [comment]
> existing Fody solutions for WPF dependent property boilerplate don't work, because AvaloniaUI rolled its property stuff.

Actually, it should be possible to implement it with custom Fody weaver. Not sure if their solution for WPF is open sourced, but if it is, it should be possible to adapt to Avalonia.

**Reply by BenMakesGames** [reply]
I took a look at a Fody package that does something similar, with thoughts of adapting it... but Fody's a lot to take in with zero experience :P but I guess so was C# Code Generators. I might take another stab at Fody...

**Comment by zxbmmmmmmmmm** [comment]
https://github.com/zxbmmmmmmmmm/PropertyGenerator

I've made a source generator for `DirectProperty` and `StyledProperty` .
Now they can be written in this way:
 (the language version must be set to `preview`)

```csharp
[GeneratedStyledProperty]
public partial int Count { get; set; }
```

```csharp
[GeneratedDirectProperty]
public partial IEnumerable? Items { get; set; } = new AvaloniaList<object>();
```

**Reply by emmauss** [reply]
This looks amazing. We've been looking at ways to implement this, but the limitations of the language was a huge road block. Now that "partial properties" is coming in the next language version, we'd love to have something similar inbuilt after it's out of preview.

**Reply by stevemonaco** [reply]
For my half-baked version, I chose to go the opposite way: the user writes the `Avalonia.RegisterXxx` part and the generator generates the property and any other bits. This has a few advantages:

1. This is much more resistant to any potential future changes with property registration.
2. It's idiomatic to Avalonia -- you're using the registration methods directly.
3. The sourcegen is MUCH easier to write because you don't need to map every single optional parameter.
4. The analyzer is MUCH easier to write because people do write those...right?
5. This prevents users from thinking they can normally get/set a dependency property. This is not the case and wipes bindings if done post-construction.
6. This prevents users from thinking they can have additional logic in getter/setter since they won't see it. This is a common footgun that almost everybody runs into at some point.

The upside to being applied to properties is that it's more idiomatic with C# properties, can inject custom logic, and maybe more refactorable? You're still going to have a code snippet either way you go.

For whatever reason, the VS navigation for Go To Definition defaults to the generated property first instead of the partial property with the attribute marker...

**Reply by LaurentInSeattle** [reply]
@stevemonaco : have you pushed your "half-baked version" to GitHub ? I am curious and would like to take a look at this alternate solution.

**Reply by robloo** [reply]
@stevemonaco your methodology where the CLR property is generated isn't the right approach I think.

The convention is to treat CLR members as the definition and then generate the rest. Everyone knows CLR properties, framework stuff should be what the generator helps with.

The most important reason for this is:

 1. Documentation is pretty important in companies and for large projects. You can't generate the XML summary docs for a CLR property... however an Avalonia properties just has boilerplate "Identifies the X property" summary.
 2. Devs need to have control over the CLR properties themselves. We need to add our own attributes and things to them, etc. Not true for Avalonia Properties.

The source generator will need to have options to control everything including default values, coercion, etc. It will be quite complicated, yes.
